### ISSUE TITLE
Some Killing Floor 1 servers have weird characters in their name.

### ISSUE THREAD

**Issue by Duckbuster** [sonicsnes/node-gamedig#121]
Hello,

I host a serverlist for some KF servers (play.scrnbalance.eu)
And I noticed some issues with some servers after they added color tags into their name.

Gamedig fails to get their name propper, it turns into Chinese characters.

Example:

```
gamedig --type killingfloor 212.109.217.192:7711 --pretty
{
  "name": "嬁瘀㔀㈀崀 ᬀ䀀䀀一漀漀戀 猀攀爀瘀攀爀 ⴀ 一椀挀攀倀愀挀欀 洀漀搀 ᬀ䠀䠄䀄嬀䌀甀爀爀攀渀琀 猀琀愀琀攀㨀 䰀伀䈀䈀夀崀",
  "map": "",
  "password": false,
  "raw": {
    "serverid": 0,
    "ip": "",
    "queryport": 0,
    "gametype": "KF-OldCity",
    "numplayers": 1667845645,
    "wavecurrent": 1886999653,
    "wavetotal": 101,
    "mutators": [
      "KillingFloorMut",
      "ServerPerksMutSE",
      "ScrnVotingHandlerMut",
      "LaSGunSkinsMut",
      "ScrnBalance",
      "NicePack"
    ],
    "rules": {
      "Server Mode": "dedicated",
      "\u001bServer Version": "\u001b1065",
      "VAC": "true",
      "Max Spectators": "32",
      "Map Voting": "true",
      "Kick Voting": "true",
      "Num trader weapons": "72",
      "ScrN Balance": "v9.61.6",
      "Max Zombies": "48",
      "": "",
      "\u001b": "\u001b"
    }
  },
  "maxplayers": 1835091813,
  "players": [],
  "bots": [],
  "connect": "212.109.217.192:7711",
  "ping": 63
}
```
As you can see, the name and maxplayers comes out bad.

Here is the manual query using the query port of the server and the results:

```
 echo -n "\\status\\" | nc -4 -u 212.109.217.192 7721 -q 1 -w 3
\gamename\redorchestra\gamever\3339\minnetver\3180\location\0\hostname\[v52] Noob server\hostport\7711\maptitle\Old City\mapname\KF-OldCity\gametype\NiceGameType\numplayers\0\maplayers\6\gamemode\openplaying\gamever\3339\minnetver\3180\password\0\queryid\1.1\final\⏎  
```

(This is what I used before I converted the site to use gamedig so it could parse it faster/more frequent.)

Any idea why this might be happening?

**Comment by Duckbuster** [comment]
Here is output of the cli with added --debug option:
```
Running attempt with options:
{ socketTimeout: 2000,
  attemptTimeout: 10000,
  maxAttempts: 1,
  port: 7712,
  protocol: 'killingfloor',
  type: 'killingfloor',
  host: '212.109.217.192',
  debug: true }
DNS Lookup: 212.109.217.192
Raw IP Address: 212.109.217.192
212.109.217.192:7712 UDP-->
Buffer length: 5 bytes
79 00 00 00 00
y

212.109.217.192:7712 <--UDP
Buffer length: 207 bytes
80 00 00 00 00 00 00 00 00 00 1f 1e 00 00 00 00 00 00 c0 01 5b
                                                            [
00 76 00 35 00 32 00 5d 00 20 00 1b 00 40 00 f8 00 40 00 4e 00
   v     5     2     ]                 @           @     N
6f 00 6f 00 62 00 20 00 73 00 65 00 72 00 76 00 65 00 72 00 20
o     o     b           s     e     r     v     e     r
00 2d 00 20 00 4e 00 69 00 63 00 65 00 50 00 61 00 63 00 6b 00
   -           N     i     c     e     P     a     c     k
20 00 6d 00 6f 00 64 00 20 00 1b 00 48 04 48 04 40 00 5b 00 43
      m     o     d                 H     H     @     [     C
00 75 00 72 00 72 00 65 00 6e 00 74 00 20 00 73 00 74 00 61 00
   u     r     r     e     n     t           s     t     a
74 00 65 00 3a 00 20 00 4c 00 4f 00 42 00 42 00 59 00 5d 00 00
t     e     :           L     O     B     B     Y     ]
00 0f 1b 40 f8 40 4b 46 2d 4f 6c 64 43 69 74 79 00 0d 4e 69 63
         @     @  K  F  -  O  l  d  C  i  t  y        N  i  c
65 47 61 6d 65 54 79 70 65 00 00 00 00 00 06 00 00 00 01 00 00
e  G  a  m  e  T  y  p  e
00 07 00 00 00 00 00 00 00 00 02 00 00 02 30 00 00 00
                                          0

Registered RTT: 65ms
UDP timeout detected
UDP timeout resolved by callback
UCS2 STRING
128 'Buffer length: 128 bytes\n01 5b 00 76 00 35 00 32 00 5d 00 20 00 1b 00 40 00 f8 00 40 00 \n   [     v     5     2     ]                 @           @     \n4e 00 6f 00 6f 00 62 00 20 00 73 00 65 00 72 00 76 00 65 00 72 \nN     o     o     b           s     e     r     v     e     r  \n00 20 00 2d 00 20 00 4e 00 69 00 63 00 65 00 50 00 61 00 63 00 \n         -           N     i     c     e     P     a     c     \n6b 00 20 00 6d 00 6f 00 64 00 20 00 1b 00 48 04 48 04 40 00 5b \nk           m     o     d                 H     H     @     [  \n00 43 00 75 00 72 00 72 00 65 00 6e 00 74 00 20 00 73 00 74 00 \n   C     u     r     r     e     n     t           s     t     \n61 00 74 00 65 00 3a 00 20 00 4c 00 4f 00 42 00 42 00 59 00 5d \na     t     e     :           L     O     B     B     Y     ]  \n00 00 \n      \n'
212.109.217.192:7712 UDP-->
Buffer length: 5 bytes
79 00 00 00 01
y

212.109.217.192:7712 <--UDP
Buffer length: 466 bytes
80 00 00 00 01 14 1b 01 01 09 1b c8 c8 c8 53 65 72 76 65 72 20
                                          S  e  r  v  e  r
4d 6f 64 65 00 12 1b 01 01 09 1b 32 dc 5a 64 65 64 69 63 61 74
M  o  d  e                       2     Z  d  e  d  i  c  a  t
65 64 00 17 1b 01 01 0a 1b c8 c8 c8 53 65 72 76 65 72 20 56 65
e  d                                S  e  r  v  e  r     V  e
72 73 69 6f 6e 00 0d 1b 01 01 0a 1b 32 dc 5a 31 30 36 35 00 0c
r  s  i  o  n                       2     Z  1  0  6  5
1b 01 01 06 1b c8 c8 c8 56 41 43 00 0d 1b 01 01 06 1b 32 dc 5a
                        V  A  C                       2     Z
74 72 75 65 00 17 1b 01 01 07 1b c8 c8 c8 4d 61 78 20 53 70 65
t  r  u  e                                M  a  x     S  p  e
63 74 61 74 6f 72 73 00 0b 1b 01 01 07 1b 32 dc 5a 33 32 00 13
c  t  a  t  o  r  s                       2     Z  3  2
1b 01 01 04 1b c8 c8 c8 4d 61 70 20 56 6f 74 69 6e 67 00 0d 1b
                        M  a  p     V  o  t  i  n  g
01 01 04 1b 32 dc 5a 74 72 75 65 00 14 1b 01 01 03 1b c8 c8 c8
            2     Z  t  r  u  e
4b 69 63 6b 20 56 6f 74 69 6e 67 00 0d 1b 01 01 03 1b 32 dc 5a
K  i  c  k     V  o  t  i  n  g                       2     Z
74 72 75 65 00 10 1b 01 01 0c 1b c8 c8 c8 4d 75 74 61 74 6f 72
t  r  u  e                                M  u  t  a  t  o  r
00 18 1b 01 01 0c 1b 01 79 c8 4b 69 6c 6c 69 6e 67 46 6c 6f 6f
                        y     K  i  l  l  i  n  g  F  l  o  o
72 4d 75 74 00 10 1b 01 01 0c 1b c8 c8 c8 4d 75 74 61 74 6f 72
r  M  u  t                                M  u  t  a  t  o  r
00 19 1b 01 01 0c 1b 01 79 c8 53 65 72 76 65 72 50 65 72 6b 73
                        y     S  e  r  v  e  r  P  e  r  k  s
4d 75 74 53 45 00 1b 1b 01 01 02 1b c8 c8 c8 4e 75 6d 20 74 72
M  u  t  S  E                                N  u  m     t  r
61 64 65 72 20 77 65 61 70 6f 6e 73 00 0b 1b 01 01 02 1b 32 dc
a  d  e  r     w  e  a  p  o  n  s                       2
5a 37 32 00 10 1b 01 01 0c 1b c8 c8 c8 4d 75 74 61 74 6f 72 00
Z  7  2                                M  u  t  a  t  o  r
1d 1b 01 01 0c 1b 01 79 c8 53 63 72 6e 56 6f 74 69 6e 67 48 61
                     y     S  c  r  n  V  o  t  i  n  g  H  a
6e 64 6c 65 72 4d 75 74 00 10 1b 01 01 0c 1b c8 c8 c8 4d 75 74
n  d  l  e  r  M  u  t                                M  u  t
61 74 6f 72 00 17 1b 01 01 0c 1b 01 79 c8 4c 61 53 47 75 6e 53
a  t  o  r                          y     L  a  S  G  u  n  S
6b 69 6e 73 4d 75 74 00 10 1b 01 01 0c 1b c8 c8 c8 4d 75 74 61
k  i  n  s  M  u  t                                M  u  t  a
74 6f 72 00 14 1b 01 01 0c 1b 01 79 c8 53 63 72 6e 42 61 6c 61
t  o  r                          y     S  c  r  n  B  a  l  a
6e 63 65 00
n  c  e

Registered RTT: 58ms
212.109.217.192:7712 <--UDP
Buffer length: 152 bytes
80 00 00 00 01 15 1b 01 01 01 1b c8 c8 c8 53 63 72 4e 20 42 61
                                          S  c  r  N     B  a
6c 61 6e 63 65 00 10 1b 01 01 01 1b 32 dc 5a 76 39 2e 36 31 2e
l  a  n  c  e                       2     Z  v  9  .  6  1  .
36 00 10 1b 01 01 0c 1b c8 c8 c8 4d 75 74 61 74 6f 72 00 11 1b
6                                M  u  t  a  t  o  r
01 01 0c 1b 01 79 c8 4e 69 63 65 50 61 63 6b 00 14 1b 01 01 08
               y     N  i  c  e  P  a  c  k
1b c8 c8 c8 4d 61 78 20 5a 6f 6d 62 69 65 73 00 0b 1b 01 01 08
            M  a  x     Z  o  m  b  i  e  s
1b 32 dc 5a 34 38 00 09 1b 01 01 0b 1b 01 01 01 00 09 1b 01 01
   2     Z  4  8
0b 1b 01 01 01 00 09 1b 01 01 0d 1b 01 01 01 00 09 1b 01 01 0d

1b 01 01 01 00


UDP timeout detected
UDP timeout resolved by callback
212.109.217.192:7712 UDP-->
Buffer length: 5 bytes
79 00 00 00 02
y

UDP timeout detected
UDP timeout resolved by callback
Size of players array: 0
Size of bots array: 0
{
  "name": "嬁瘀㔀㈀崀 ᬀ䀀䀀一漀漀戀 猀攀爀瘀攀爀 ⴀ 一椀挀攀倀愀挀欀 洀漀搀 ᬀ䠀䠄䀄嬀䌀甀爀爀攀渀琀 猀琀愀琀攀㨀 䰀伀䈀䈀夀崀",
  "map": "",
  "password": false,
  "raw": {
    "serverid": 0,
    "ip": "",
    "queryport": 0,
    "gametype": "KF-OldCity",
    "numplayers": 1667845645,
    "wavecurrent": 1886999653,
    "wavetotal": 101,
    "mutators": [
      "KillingFloorMut",
      "ServerPerksMutSE",
      "ScrnVotingHandlerMut",
      "LaSGunSkinsMut",
      "ScrnBalance",
      "NicePack"
    ],
    "rules": {
      "Server Mode": "dedicated",
      "\u001bServer Version": "\u001b1065",
      "VAC": "true",
      "Max Spectators": "32",
      "Map Voting": "true",
      "Kick Voting": "true",
      "Num trader weapons": "72",
      "ScrN Balance": "v9.61.6",
      "Max Zombies": "48",
      "": "",
      "\u001b": "\u001b"
    }
  },
  "maxplayers": 1835091813,
  "players": [],
  "bots": [],
  "connect": "212.109.217.192:7711",
  "ping": 58
}
```

**Comment by Duckbuster** [comment]
Any news?

**Comment by sonicsnes** [comment]
I've cleaned up some of the unreal ucs2 decoding in the latest release, but was unable to reproduce your issue exactly. Seems like querying the server works properly now 🤔

Please re-open if you experience this again.

**Comment by Duckbuster** [comment]
> I've cleaned up some of the unreal ucs2 decoding in the latest release, but was unable to reproduce your issue exactly. Seems like querying the server works properly now 

I still have the issue (after updating to the newest version)
Also let my friend have a try at it, he has the same issue.
Make sure you test at the server I linked, that one has the issue:
gamedig --type killingfloor 212.109.217.192:7711 --pretty

> Please re-open if you experience this again.

I can't re-open it since you closed it :-(

**Comment by sonicsnes** [comment]
I'm not 100% sure why this is happening, but that specific kf server is inserting a random extra byte between the length and the data of the pascal ucs-2 string for the server name. You can temporarily fix it by updating to the latest version of gamedig, and in protocols/unreal2.js, add:
```
            const peek = reader.uint(1);
            if (peek !== 1) reader.skip(-1);
```
immediately after
```
            length = (length&0x7f)*2;
```

**Comment by Duckbuster** [comment]
I think it's because they insert some special characters to make the server name have colors inside the in-game server browser.

I will check it out when I can. Thanks

**Comment by Duckbuster** [comment]
Thanks, that fixes it.
Will this be added in officially in the next version?
Or should I keep manually modifying it for future updates?

**Comment by sonicsnes** [comment]
I'm a bit reluctant to merge, because I really have no explanation for why this is happening in only some killing floor servers.

I suppose it's extraordinarily rare for a server to legitimately have a 1 in the least significant byte of the first ucs-2 character, so it's probably safe enough.

**Comment by sonicsnes** [comment]
Fixed in 2.0.17. Also squashed the bug that was leaving random lingering `\u001b`s in your server rules.

**Comment by Duckbuster** [comment]
Thank you!
